### ISSUE TITLE
CA1814 is reported in generated code, for a call to a method I don't control

### ISSUE THREAD

**Issue by thomaslevesque** [dotnet/roslyn-analyzers#3487]
#### Analyzer package

[Microsoft.CodeAnalysis.FxCopAnalyzers](https://www.nuget.org/packages/Microsoft.CodeAnalysis.FxCopAnalyzers)

#### Package Version

v2.9.8

#### Diagnostic ID

Example: [CA1814](https://docs.microsoft.com/en-us/visualstudio/code-quality/ca1814)

#### Repro steps

1. In EF Core model configuration, define seed data for an entity using `builder.Entity<MyEntity>().HasData(new MyEntity { ... }, new MyEntity { ... } ... )`
2. Generate a new migration (`dotnet ef migrations add`)

#### Expected behavior

No warning

#### Actual behavior

Warning CA1814 on the argument for `values`:

```csharp
            migrationBuilder.InsertData(
                table: "MyEntity",
                columns: new[] { "Id", "Name" },
                values: new object[,]
                {
                    { 1, "Foo" },
                    { 2, "Bar" }
                });
```

`InsertData` is a method from EF Core, I don't control it. I can't do anything about its signature, so the warning is useless in this case ; the only way to "fix" it would be to not call this method...

**Comment by Evangelink** [comment]
The rule analyze object creations but that could produce too many FPs:
1. object creation as argument to a method call (see previous comment)
2. object creation as "argument" to the return. If the method overrides a base method or implement an interface we have to use a multi-dimensional object.
3. object creation assigned to a local variable. This is a little more tricky but that's quite likely this variable is going to be passed to a method or returned so if we can't fix the return or method call then having to suppress the issue for the variable is also a pain.

For these reasons I think the object creation part should be removed. As the rule already exists, we might want to have a new user option `analyze_object_creation` turned off by default to reduce false-positives while allowing to keep the old behavior.

@mavasani WDYT?

**Comment by mavasani** [comment]
@thomaslevesque Please consider turning off this rule for your project or adding a source suppression for the false positives.

@Evangelink We should consider turning off this rule by default - this has already been identified as a potential candidate for removal in NetAnalyzers package, so would only exist for legacy compat in case someone really wants this rule for their project and are fine to live with the false positives: https://github.com/dotnet/roslyn-analyzers/blob/f15404b312295e7cee16fb40b4c8d3f91f10f087/src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/QualityGuidelines/PreferJaggedArraysOverMultidimensional.cs#L31

**Comment by Evangelink** [comment]
I saw that the rule was marked as `CandidateForRemoval` and I was actually curious on the why, do you have any info? As far as I know the performance hit is still present in netcore so it still seems to be a valuable rule to me.

**Comment by thomaslevesque** [comment]
> @thomaslevesque Please consider turning off this rule for your project or adding a source suppression for the false positives.

Yup, that's what I did

**Comment by Evangelink** [comment]
> I saw that the rule was marked as `CandidateForRemoval` and I was actually curious on the why, do you have any info? As far as I know the performance hit is still present in netcore so it still seems to be a valuable rule to me.

@mavasani Ping

**Comment by Evangelink** [comment]
As far as I can tell the rule is now disabled by default so there is not much to do for this ticket.
Tagging @mavasani 

I would still be interested to know why we want to get rid of this rule.

**Comment by Evangelink** [comment]
Ping @mavasani

**Comment by mavasani** [comment]
> I saw that the rule was marked as CandidateForRemoval and I was actually curious on the why, do you have any info? As far as I know the performance hit is still present in netcore so it still seems to be a valuable rule to me.

@sharwell, I don't see the reason for marking this as a candidate rule for removal in our triage meeting notes. Do you happen to recall the reason?

**Comment by sharwell** [comment]
If a multi-dimensional array is appropriate for the problem set, then it's appropriate for the problem set. This rule is unlikely to make meaningful progress against a performance issue related to sparse arrays. As a rule with no clear merits or driving need, there is no current path forward where we would suggest any user enable it.

**Comment by Evangelink** [comment]
That's interesting, I think it'd be nice to share this info in some .net news because I think a lot of people have been taught not to use multi-dim for performance consideration.

To back your point it seems that some benchmarks actually show the multi-dim array being faster than jagged array on .net core.

@mavasani I would suggest to deprecate the rule in the doc, close all associated tickets as won't fix and remove the rule in the branch `6.0.1xx-preview1`. WDYT?

**Comment by mavasani** [comment]
>  I would suggest to deprecate the rule in the doc, close all associated tickets as won't fix and remove the rule in the branch 6.0.1xx-preview1. WDYT?

Tagging @jmarolf - I think we should bring up deprecation of old CA rules in one of the netanalyzers v-team design meetings before deciding to deprecate it.

**Comment by jmarolf** [comment]
I think this is a good thing to discuss.
